### PR TITLE
WIP: Add unified descriptor schema

### PR DESCRIPTION
**Pull request by OtonicarJan**
## Checklist

* [ ] Update CHANGELOG.rst for each commit separately:
  * Pay attention to write entries under the "Unreleased" section.
  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
    before non-breaking changes.
  * If a commit modifies a feature listed under "Unreleased" section,
    it might be sufficient to modify the existing CHANGELOG entry from previous
    commit(s).
* [ ] Bump the process version:
  * **MAJOR version (first number)**: Backward incompatible changes (changes
    that break the api/interface). Examples: renaming the input/output, adding
    mandatory input, removing input/output...
  * **MINOR version (middle number)**: add functionality or changes in a
    backwards-compatible manner. Examples: add output field, add non-mandatory
    input parameter, use a different tool that produces same results...
  * **PATCH version (last number)**: changes/bug fixes that do not affect
    the api/interface. Examples: typo fix, change/add warning messages...
* [ ] All inputs are used in process.
* [ ] All output fields have a value assigned to them.

### Processes ###

* Code that was once accepted should be kept like it is. It should be changed
  only when its functionality has changed. Small parts may be changed
  to fix bugs (by changing code only to make it nicer you might introduce a bug).
* Set the optimal number of cores that the tool can use (check if parallelization
  can be performed).
* Description of the tool should be self explanatory with links to the tool's
  homepage and publication.
* Write explanatory input labels and add short command line parameter names
  in square brackets (for example: Maximum length [``--maxlength``]).
* Where possible provide default values for the process input fields.
  If they are provided, then ``required = False`` is not needed.
* Add descriptive error and warning messages.
* Use ``Cmd()`` for each call in the process (call of a tool, a
  Bash command, a Python script or an R script) and handle return codes.
* Add ``self.progress()`` calls where appropriate.

### Commit and PR messages ###

* Each commit should be minimal (i.e. 1 change) and self-contained (including
  tests).
* Mark incomplete PRs with the [WIP] (Work In Progress) tag.
* Do not paste links to private repositories from the public one.

### Tests ###

* Add meaningful names of test files:
  * Genomes: ``genome_species.fa.gz``
  * Annotations: ``annotation_species.gtf.gz``
  * Adapters: ``adapters-name/type-source``.
* Specific files should be saved in a folder with the name of the tested tool.
  This folder should include separate subfolders for inputs and outputs.
* Keep test files' size small. If there is no other option, add the large
  test file to the ``files/large`` directory so it will be tracked with Git LFS.
* Pay attention to edge cases: what will happen if any data is
  missing, if the length of some list is 0, if a string is given where one
  expects an integer, if a file is empty, ...
* When testing workflows, try to verify workflow success by checking one
  of the last steps. Pick the one that is most likely to fail. Also test
  that all steps have 'OK' status.

### CODE REVIEW

File: resolwe_bio/descriptors/unified-descriptor.yml [hunk 1]
```diff
@@ -0,0 +1,1877 @@
...
+            - label: RNA-seq
+              value: rna-seq
+            - label: DNA-seq
+              value: dna-seq
+            - label: Nanostring
+              value: nanostring
+            - label: Microarray
+              value: microarray
+            - label: Edge-seq
+              value: edge-seq
```
Comment (OtonicarJan) [inline resolwe_bio/descriptors/unified-descriptor.yml:773]: What to do with this field? Field `platform` in `experiment` group above is not the same as this one.

File: resolwe_bio/descriptors/unified-descriptor.yml [hunk 2]
```diff
@@ -0,0 +1,1877 @@
...
+        - name: characteristics
+          label: Characteristics
+          type: basic:string
+          required: false
+        - name: growth_protocol
+          label: Growth Protocol
+          type: basic:string
+          description: |
+            Describe the conditions that were used to grow or maintain cells prior to extract preparation.
+          required: false
```
Comment (OtonicarJan) [inline resolwe_bio/descriptors/unified-descriptor.yml:808]: Should this be seperated from `growth_protocol` in the group `general`?

File: resolwe_bio/descriptors/unified-descriptor.yml [hunk 3]
```diff
@@ -0,0 +1,1877 @@
...
+        - name: freezing_protocol
+          label: Freezing Protocol
+          type: basic_string
+          required: false
+        - name: treatment_protocol
+          label: Treatment Protocol
+          type: basic:string
+          description: |
+            Describe how the sample was treated, at different timepoints, if using non-chemical treatments e.g. exercise, aging, etc.
+          required: false
```
Comment (OtonicarJan) [inline resolwe_bio/descriptors/unified-descriptor.yml:824]: Should this be separated from `treatment_protocol` in the group `general`?

File: resolwe_bio/descriptors/unified-descriptor.yml [hunk 4]
```diff
@@ -0,0 +1,1877 @@
...
+            - label: Pass
+              value: PASS
+            - label: Warning
+              value: WARNING
+            - label: Fail
+              value: FAIL
+        - name: message
+          label: Message
+          type: basic:string
+          required: false
```
Comment (OtonicarJan) [inline resolwe_bio/descriptors/unified-descriptor.yml:273]: These two fields were part of the nested group `qc` before. Prefix "quality" was added to "status". What about message?

File: resolwe_bio/descriptors/unified-descriptor.yml [hunk 5]
```diff
@@ -0,0 +1,1877 @@
...
+        - name: antibody_catalog_number
+          label: Antibody catalog number
+          type: basic:string
+          required: false
+          hidden: "experiment.assay_type != 'chip-seq' && experiment.assay_type != 'chipmentation'"
+        - name: antibody_lot_number
+          label: Antibody lot number
+          type: basic:string
+          required: false
+          hidden: "experiment.assay_type != 'chip-seq' && experiment.assay_type != 'chipmentation'"
```
Comment (OtonicarJan) [inline resolwe_bio/descriptors/unified-descriptor.yml:691]: These fields were part of the nested group `chip_seq` before. Should we add some prefix/suffix to the name and label to make it more clear?

**Review by OtonicarJan — COMMENTED** [review]
@jkokosar @romunov @uksurd88

File: resolwe_bio/descriptors/unified-descriptor.yml (hunk 2 above)
Comment (uksurd88) [reply resolwe_bio/descriptors/unified-descriptor.yml:814]: @OtonicarJan This should be moved to the group of `cell_line_information` as in https://github.com/genialis/genialis-bio/blob/master/genialis_bio/inhibrx/descriptors/Inhibrx-cell-line.yml

File: resolwe_bio/descriptors/unified-descriptor.yml [hunk 6]
```diff
@@ -0,0 +1,1537 @@
...
+        #     - label: Fail
+        #       value: FAIL
+        # - name: quality_message
+        #   label: Quality message
+        #   type: basic:string
+        #   required: false
+        - name: cell_line
+          label: Cell line information
+          group:
+            # - name: primary_cell_line
```
Comment (uksurd88) [inline resolwe_bio/descriptors/unified-descriptor.yml:84]: we can enable primary_cell_line again if required, for now we commented it out. Rest of the fields should be sufficient to carry most of the cell line information 
cc @amiamrita

File: resolwe_bio/descriptors/unified-descriptor.yml [hunk 7]
```diff
@@ -0,0 +1,1537 @@
...
+              required: false
+            - name: batch
+              label: Batch
+              type: basic:integer
+              required: false
+            - name: group
+              label: Group
+              type: basic:string
+              required: false
+            - name: sex
```
Comment (uksurd88) [inline resolwe_bio/descriptors/unified-descriptor.yml:170]: we choose `sex` over `gender`.

File: resolwe_bio/descriptors/unified-descriptor.yml [hunk 8]
```diff
@@ -0,0 +1,1537 @@
...
+                  value: nextseq_500
+                - label: HiSeq 2500
+                  value: hiseq_2500
+                - label: HiSeq 2000
+                  value: hiseq_2000
+                - label: NovaSeq 6000
+                  value: novaseq_6000
+                - label: Other
+                  value: other
+            # - name: target_of_chip_assay
```
Comment (uksurd88) [inline resolwe_bio/descriptors/unified-descriptor.yml:1477]: We disabled `ChIP-seq` but can enabled later. 
@jkokosar

**Review by uksurd88 — COMMENTED** [review]
Good work @OtonicarJan, please see one comment from me, rest all good.

File: resolwe_bio/descriptors/unified-descriptor.yml (hunk 8 above)
Comment (OtonicarJan) [reply resolwe_bio/descriptors/unified-descriptor.yml]: ChIP-seq fields were removed. Changelog was updated accordingly.

File: resolwe_bio/descriptors/unified-descriptor.yml (hunk 6 above)
Comment (OtonicarJan) [reply resolwe_bio/descriptors/unified-descriptor.yml]: Field `primary_cell_line` was removed and can be added in the future if needed. Changelog was updated accordingly.

File: resolwe_bio/descriptors/unified-descriptor.yml [hunk 9]
```diff
@@ -0,0 +1,1448 @@
...
+          type: basic:integer
+          required: false
+        - name: collection
+          label: Collection
+          type: basic:string
+          required: false
+        - name: species
+          label: Species
+          type: basic:string
+          default: Homo sapiens
```
Comment (jkokosar) [inline resolwe_bio/descriptors/unified-descriptor.yml:28]: It wrong to set the default value to `Homo sapiens`. If this schema is to be the default for all new samples, all non-human samples will be wrongfully annotated by default.

File: resolwe_bio/descriptors/unified-descriptor.yml [hunk 10]
```diff
@@ -0,0 +1,1448 @@
...
+              value: Dictyostelium purpureum
+            - label: Drosophila melanogaster
+              value: Drosophila melanogaster
+            - label: Homo sapiens
+              value: Homo sapiens
+            - label: Macaca mulatta
+              value: Macaca mulatta
+            - label: Mus musculus
+              value: Mus musculus
+            - label: Odocoileus virginianus texanus
```
Comment (jkokosar) [inline resolwe_bio/descriptors/unified-descriptor.yml:47]: Remove `Odocoileus virginianus texanus` and `Solanum tuberosum` from the choices menu.

File: resolwe_bio/descriptors/unified-descriptor.yml [hunk 11]
```diff
@@ -0,0 +1,1448 @@
...
+              type: basic:integer
+              required: false
+            - name: mesenchymal
+              label: Mesenchymal
+              type: basic:string
+              description: |
+                Select if the cells are of mesenchymal type.
+              required: false
+              choices:
+                - label: "Yes"
```
Comment (jkokosar) [inline resolwe_bio/descriptors/unified-descriptor.yml:319]: For all cases of "Yes" / "No", please check if the quotes are required. It suspect they are redundant.

File: resolwe_bio/descriptors/unified-descriptor.yml (hunk 4 above)
Comment (jkokosar) [reply resolwe_bio/descriptors/unified-descriptor.yml]: @OtonicarJan @uksurd88 Removal of the `qc` block (and renaming of the fields therein) is a regression in regards to the existing FE functionality. I vote for this to remain available in the `general` schema, as it allows for manual and possibly automated  marking of the sample QC status in the future.

File: resolwe_bio/descriptors/unified-descriptor.yml (hunk 11 above)
Comment (OtonicarJan) [reply resolwe_bio/descriptors/unified-descriptor.yml:319]: Tried running `test_api.py` with modified `general-clinical` DS (yes and no without quotes), but I get this error
```
RuntimeError:     VALIDATION ERROR: General clinical annotation
        path:       deque(['choices', 0, 'label'])
        message:    True is not of type 'string', 'integer', 'number'
        validator:  type
        val. value: ['string', 'integer', 'number']
```

File: resolwe_bio/descriptors/unified-descriptor.yml [hunk 12]
```diff
@@ -0,0 +1,1445 @@
...
+              label: Sampling time
+              type: basic:string
+              required: false
+              choices:
+                - label: Pretreatment
+                  value: pretreatment
+                - label: Posttreatment
+                  value: posttreatment
+            - name: tumor
+              label: Tumor
```
Comment (romunov) [inline resolwe_bio/descriptors/unified-descriptor.yml:412]: Perhaps "Tumor source"?

File: resolwe_bio/descriptors/unified-descriptor.yml [hunk 13]
```diff
@@ -0,0 +1,1445 @@
...
+              description: |
+                Select what molecule your samples are from.
+              required: false
+              choices:
+                - label: RNA
+                  value: rna
+                - label: DNA
+                  value: dna
+            - name: normal
+              label: Normal
```
Comment (romunov) [inline resolwe_bio/descriptors/unified-descriptor.yml:423]: Perhaps "Normal source"?

File: resolwe_bio/descriptors/unified-descriptor.yml [hunk 14]
```diff
@@ -0,0 +1,1445 @@
...
+              required: false
+            - name: escalation_dose
+              label: Escalation dose
+              type: basic:string
+              required: false
+            - name: subject_status
+              label: Subject status
+              type: basic:string
+              description: |
+                Specify if the subject is currently on therapy or not.
```
Comment (romunov) [inline resolwe_bio/descriptors/unified-descriptor.yml:565]: Could this field be boolean?

File: resolwe_bio/descriptors/unified-descriptor.yml [hunk 15]
```diff
@@ -0,0 +1,1445 @@
...
+                  value: pd
+            - name: pfs_date
+              label: PFS date
+              type: basic:date
+              description: |
+                Monitoring Progress Free Survival since.
+              required: false
+            - name: pfs
+              label: PFS
+              type: basic:decimal
```
Comment (romunov) [inline resolwe_bio/descriptors/unified-descriptor.yml:785]: Since we're storing this in months, I would assume we're recording integers.

File: resolwe_bio/descriptors/unified-descriptor.yml [hunk 16]
```diff
@@ -0,0 +1,1445 @@
...
+              description: |
+                Programmed death-ligand 1 positive tumor microenvironment cell count.
+              required: false
+            - name: pah_score
+              label: PAH score
+              type: basic:decimal
+              description: |
+                Pulmonary Arterial Hypertension score.
+              required: false
+            - name: p53_score
```
Comment (romunov) [inline resolwe_bio/descriptors/unified-descriptor.yml:1191]: Perhaps this one could use a description of what p53 score is.

File: resolwe_bio/descriptors/unified-descriptor.yml [hunk 17]
```diff
@@ -0,0 +1,1445 @@
...
+                Combined positive score.
+              required: false
+            - name: description_of_biomarkers
+              label: Description of biomarkers
+              type: basic:string
+              required: false
+        - name: tme_panel_1
+          label: TME Panel-1
+          group:
+            - name: ann_call
```
Comment (romunov) [inline resolwe_bio/descriptors/unified-descriptor.yml:1240]: If classes are known, perhaps this could be a drop-down list?

File: resolwe_bio/descriptors/unified-descriptor.yml [hunk 18]
```diff
@@ -0,0 +1,1445 @@
...
+              type: basic:decimal
+              description: |
+                Probability immune-desert.
+            - name: ann_prob_is
+              label: ANN Prob IS
+              type: basic:decimal
+              description: |
+                Probability immune-suppressed.
+            - name: description_tme_panel_1
+              label: Description TME Panel-1
```
Comment (romunov) [inline resolwe_bio/descriptors/unified-descriptor.yml:1285]: Description _of_ TME Panel-1?

File: resolwe_bio/descriptors/unified-descriptor.yml [hunk 19]
```diff
@@ -0,0 +1,1445 @@
...
+              label: Extract protocol
+              type: basic:string
+              description: |
+                Describe the protocols used to extract and prepare the material to be sequenced.
+              required: false
+            - name: molecule
+              label: Molecule
+              type: basic:string
+              description: |
+                Select what molecule your samples are from, e.g. totalRNA, polyA RNA, genomic DNA, etc.
```
Comment (romunov) [inline resolwe_bio/descriptors/unified-descriptor.yml:1375]: Add space between total and RNA.

**Review by uksurd88 — APPROVED** [review]

**Review by jkokosar — APPROVED** [review]
